File: 13 лаба/13B.py

fin = open("search2.in",  "r")
fout = open("search2.out", "w")

pattern = list(fin.readline())
pattern.remove('\n')
text = list(fin.readline())

M = len(pattern)
N = len(text)

Hash_pattern = 0

k=31
m = k**M
mod = 1e3+7

for c in pattern:
	x = (ord(c) - ord('a')+1)
	Hash_pattern = (Hash_pattern*k + x) % mod


Hash = [0 for i in range(N-M+1)]

for c in text[:M]:
	x = (ord(c) - ord('a')+1)
	Hash[0] = (Hash[0]*k + x) % mod


for i in range(N-M):
	Hash[i+1] = (Hash[i]*k - (m*(ord(text[i])-ord('a')+1)) + (ord(text[i+M])-ord('a')+1)) % mod

places = []
count = 0

for i in range(N-M+1):
	if Hash[i]!=Hash_pattern:
		continue
	check = 1
	for j in range(M):
		if text[i+j]!=pattern[j]:
			check = 0
			break
	if check:
		places.append(i+1)
		count+=1

fout.write(str(count))
fout.write("\n")
for i in range(len(places)): 
	fout.write(str(places[i]))
	fout.write(" ")

fout.write("\n")
for i in range(len(Hash)): 
	fout.write(str(Hash[i]))
	fout.write(" ")

fout.write("\n")
fout.write(str(ord('b') - ord('a')))
fin.close()
fout.close()

File: 13 лаба/13B.cpp
#include <fstream>
#include <vector>
#include <cmath>


using namespace std;

int main() {

    ifstream fin("search2.in");
    ofstream fout("search2.out");

    string text, pattern;
    int count = 0;
    bool check;


    fin >> pattern >> text;

    vector<int> places;

    int M = pattern.length();
    int N = text.length();

    int Hash_pattern = 0;
    int x = 3;
    long long x_m = pow(x,M);
    int p = 1007;
    vector<int> Hash(N-M+1, 0);

    for (int i=0; i<M;i++){
        Hash_pattern = (Hash_pattern*x + (pattern[i]-'a')) % p;
    }

    for(int i=0;i<M;i++){
        Hash[0] = (Hash[0]*x + (text[i]-'a')) % p;
    }

    for(int i=0; i<(N-M+1);i++){
        Hash[i+1] = (Hash[i]*x - (x_m*(text[i]-'a')) + (text[i+M]-'a')) % p;

        /*for (int i=0; i< Hash.size(); i++){
        fout << Hash[i] << " ";
    }
    fout << endl;*/
    }

    //fout << Hash_pattern << endl;

    for (int i=0; i<N-M+1;i++){
        if (Hash[i]!=Hash_pattern){
            continue;
        }
        check = true;

        for(int j=0; j<M;j++){
            if(text[i+j]!=pattern[j]){
                check = false;
                break;
            }
        }
        if (check){
            places.push_back(i+1);
            count++;
        }
    }

    fout << count << endl;
    for (int i=0; i< places.size(); i++){
        fout << places[i] << " ";
    }

    return 0;
}
